simulator/cube.py

```diff
@@ -0,0 +1,401 @@
+
+  def disconnect(self, side):
+    """ Removes a connection from the cube.
+    Args:
+      side: The side to disconnect. """
+    if self.__connected[side] is None:
+      return
+
+    self.__connected[side] = None
+    # Report a state change.
+    self.__config_changed_hook()
+
+  def get_pos(self):
+    """
+    Returns:
+      The current position of the cube as (x, y). """
+    case = self.__cube_shapes[0]
+    return case.get_pos()
+
+  def set_pos(self, x, y):
+    """ Sets the position of the cube.
+    Args:
+      x: The new x position.
+      y: The new y position. """
+    # Figure out the offset from the old position.
+    old_x, old_y = self.get_pos()
+    move_x = x - old_x
+    move_y = y - old_y
+
+    for shape in self.__cube_shapes:
+      shape.move(move_x, move_y)
+
+    # Update the canvas.
+    self.__canvas.update()
+
+  def get_display(self):
+    """
+    Returns:
+      The display object for this cube. """
+    return self.__screen
+
+  def run_app(self, app):
+    """ Run a new application on the cube.
+    Args:
+      app: The class of the app to run. """
+    self.__application = app
+    self.__application.run(self)
+
+  def drag(self, event):
+    """ Respond to a mouse drag while the cube is selected.
+    Args:
+      event: The drag event to respond to. """
+    if not self.__dragging:
+      # We're not actively dragging this cube. Don't move it.
+      return
+
+    # Move the entire cube.
+    new_x, new_y = event.get_pos()
+    # Figure out how much to move it.
+    move_x = new_x - self.__prev_mouse_x
+    move_y = new_y - self.__prev_mouse_y
+    for shape in self.__cube_shapes:
+      shape.move(move_x, move_y)
+
+    self.__prev_mouse_x = new_x
+    self.__prev_mouse_y = new_y
+
+  def clear_drag(self):
+    """ Clears the current dragging state. """
+    if not self.__dragging:
+      # Not dragging this cube.
+      return
+
+    self.__dragging = False
+    assert Cube._selected == self
+    Cube._selected = None
+
+  def send_message(self, side, message):
+    """ Sends a message to a cube directly connected to this one.
+    Args:
+      side: The side that the recipient is connected on.
+      message: The string message to send. """
+    other = self.__connected[side]
+    if other is None:
+      # Make sure there's something there.
+      raise ValueError("No cube connected on side %s'." % (side))
+
+    # Pass the message to the cube.
+    other_side = Cube.Sides.opposite(side)
+    other.receive_message(other_side, message)
+
+  def receive_message(self, side, message):
+    """ Receives a message from a connected cube.
+    Args:
+      side: The side that the sender is connected to.
+      message: The string message being received. """
+    if not self.__application:
+      # With no app, the message gets dropped.
+      return
+
+    # Otherwise, pass it to the app.
+    self.__application.on_message_receive(side, message)
+
+  def is_near(self, other, threshold=100):
```

Comment:
I think we should probably get rid of this method since we don't really use it anymore.

Reply:
I was going to get rid of this method, but remembered you mentioning that the cubes will have the capability to detect nearby but not touching cubes. Is it definitely irrelevant?